modules/trainer/GenericTrainer.py

```diff
@@ -319,6 +322,63 @@ def __sample_during_training(
 
         torch_gc()
 
+    def __validate(self, train_progress):
+        self.validation_data_loader.get_data_set().start_next_epoch()
+        current_epoch_length_validation = self.validation_data_loader.get_data_set().approximate_length()
+        if current_epoch_length_validation > 0 and self.__needs_validate(train_progress):
+            torch_gc()
+
+            step_tqdm_validation = tqdm(
+                self.validation_data_loader.get_data_loader(),
+                desc="validation_step",
+                total=current_epoch_length_validation,
+                initial=0)
+
+            accumulated_loss_per_concept = {}
+            concept_counts = {}
+            mapping_seed_to_label = {}
+            mapping_label_to_seed = {}
+
+            for epoch_step_validation, validation_batch in enumerate(step_tqdm_validation):
+
+                if self.__needs_gc(train_progress):
+                    torch_gc()
+
+                with torch.no_grad():
+                    model_output_data = self.model_setup.predict(self.model, validation_batch, self.config,
+                                                                 train_progress)
+                    loss_validation = self.model_setup.calculate_loss(self.model, validation_batch, model_output_data,
+                                                                      self.config, mean=False)
+
+                for concept_name, concept_seed, image_path, loss in zip(validation_batch["concept_name"],
+                                                                        validation_batch["concept_seed"],
+                                                                        validation_batch["image_path"],
+                                                                        loss_validation):
+                    concept_seed = concept_seed.item()
+
+                    label = concept_name if concept_name else os.path.basename(os.path.dirname(image_path))
+                    # check and fix collision to display both graphs in tensorboard
```

Comment:
I'm working on another pull request that will, as a side-effect, force concept names to be unique. As a result you can just assume names are unique and not worry about this.